snapshot/snapshot_computation.py

```diff
@@ -0,0 +1,305 @@
+#!/usr/bin/env python3
+"""
+Snapshot Computation is a tool to initialize micro simulations and create a snapshot database by simulating
+micro simulations related to a set of prescribed macro parameters.
+This files the class SnapshotComputation which has the following callable public methods:
+
+- solve
+
+This file is directly executable as it consists of a main() function. Upon execution, an object of the class SnapshotComputation is created using a given JSON file,
+ the solve method is called.
+
+Detailed documentation: https://precice.org/tooling-micro-manager-overview.html
+"""
+
+import argparse
+import importlib
+import os
+import sys
+from mpi4py import MPI
+import numpy as np
+import logging
+import time
+
+from .config_snapshot import SnapshotConfig
+from .read_write_hdf import ReadWriteHDF as hp
+from micro_manager.micro_simulation import create_simulation_class
+
+sys.path.append(os.getcwd())
+
+
+class SnapshotComputation:
```

Comment:
Can we make this a child class of the MicroManager class? We redefine many member variables anyway. Perhaps this would make the overall implementation of the snapshot computation cleaner. The justification for doing this is that the snapshot computation is in a way a use case of the Micro Manager, but for a different purpose. So we are in fact running the Micro Manager, but just in a different way.

Reply:
I thought about inheritance but to me, there is too much difference in all the class methods to make it seem sensible. I believe it's possible to inherit from the Micro Manager but the Micro Manager also sets many member variables that are not required or available for the snapshot computation, e.g. everything related to preCICE, so I'm not sure how much of the Micro Manager could be inherited. Not using preCICE in the snapshot computation makes it difficult to reuse parts of the Micro Manager as much of the managing process depends on information from or consists of information written to preCICE. What could make it more natural to integrate the Micro Manager into the snapshot computation is to have a class that can imitate preCICE interface by performing the necessary operations for the snapshot computation and handling the snapshot database writing. This class could be used for the snapshot computation and communicate macro parameters to the snapshot computation. This approach could allow us to reuse much of the Micro Manager for the snapshot computation.

Reply:
> there is too much difference in all the class methods to make it seem sensible

If a function is redefined in the child class, the original function in the parent class will be ignored. So we will be replacing the functionality in a way.

Are you suggesting we mock the preCICE interface and have a dummy Micro Manager class? How would the original `MicroManager` class then function? Does it also inherit from this dummy class?

Reply:
I thought that maybe it is possible to set `self._participant` in the snapshot child of the Micro Manager as a class within the Micro Manager package that replaces preCICE in the snapshot computation by having the necessary methods of preCICE and e.g. `read_data`  could read parameter data from the parameter file and `write_data` could write output to a file. Then it could be possible to use the Micro Manager class for most of the snapshot computation. But crash handling and domain decomposition must be handled differently and would still be complicated.

If that is not possible, I think it is possible to use the `__init__` from the Micro Manager but that `initialize()` and `solve()` are too different to use the corresponding methods from the Micro Manager.

Reply:
This sounds rather complicated and over-engineered. I would suggest inheriting from the Micro Manager class, using the member variables which are relevant to the snapshot computation, and redefining the necessary functions once more. When you redefine `initialize()` and `solve()` in the `SnapshotComputation` class, the original functions will be ignored.

Reply:
Okay, that sounds like a good compromise to include inheritance to me! I know they will be ignored. I thought it might be possible to use much of the entire Micro Manager for the snapshot computation, but I agree it would be over-engineered that way and I am not convinced it would be possible in a clean way.

Reply:
Sounds good :+1: we will at least shorten the constructor and such. After we have merged the feature, we can think again about redesigning the functions in `SnapshotComputation` to somehow use functions from `MicroManager`.

Reply:
After looking further into this, inheriting from `MicroManager` looks complicated (I probably saw these issues the first time I thought about inheritance). Member variables set in the constructor are only available if the constructor is called, but calling the `MicroManager` constructor will lead to errors (and warnings) as the `MicroManager` constructor tries to read a Micro Manager-type config file that does not coincide with the snapshot computation file and it tries to create a preCICE participant, which cannot work. I cannot find any clean way around this. So the only option(s) for inheritance I see is to let both the Micro Manager and the Snapshot Computation inherit from a common base class that handles common member variables (or re-designing the Micro Manager constructor).